massaging vm quickstart (#118)
cisco-open/camblet: site/docs/src/content/docs/start/kubernetes.mdx

sudo chmod 644 /etc/camblet/kubelet-ca.crt
```

> The agent configuration resides in /etc/camblet/config.yaml, modify it the enable the Kubernetes metadata collector.

Reviewer: ```suggestion
The agent configuration resides in `/etc/camblet/config.yaml`, modify it the enable the Kubernetes metadata collector.
```

      enabled: false
```

The Camblet agent must be restarted after the configuration change.

```sh
sudo systemctl restart camblet.service
```

Let's check if the agent is indeed able to collect metadata from Kubernetes for a process that runs within a pod.
K3s comes with Traefik installed so the following command can be used as a test.

```sh

> $ camblet agent augment $(pidof traefik)

Reviewer: ```suggestion
camblet agent augment $(pidof traefik)
```

Inside the Alpine container, first, we have to install cURL:

```sh
apk add curl
```

Next, try to connect to the echo server on echo:

```sh
curl echo
```

> > A second try could be necessary to achive the wanted effect!

Reviewer: ```suggestion
> A second attempt may be necessary to achieve the desired effect.
```

The output should be similar:

```sh
curl: (56) Recv failure: Connection reset by peer
```

The connection failed, as it was supposed to. Since Camblet now protects the echo-server workload with mTLS nothing can communicate with it without a trusted client certificate.

Now exit from the Alpine container and let's create a policy for cURL as well. A running process of cURL is the easiest way to create a new policy for cURL with the Camblet CLI. Since cURL is not running continuously like the server, a "dummy" command is needed to force it to run until generate the policy.

The following command will run a cURL process for 30 seconds, the 1.2.3.4 IP address is non-routed and the connection to it going to time out. This is plenty of time to generate the policy.

```sh
kubectl exec alpine -- sh -c "curl --connect-timeout 30 1.2.3.4 >/dev/null 2>&1 &"
```

Let's generate a policy for cURL

```sh

> camblet --config /etc/camblet/config.yaml agent generate-policy $(pidof curl) curl | sudo tee /etc/camblet/policies/curl.yaml

Reviewer: ```suggestion
camblet agent generate-policy $(pidof curl) curl | sudo tee /etc/camblet/policies/curl.yaml
```

Services has to be registered into the [service registry](/docs/concepts/service-registry) through service definitions.

**Sample Service discovery configuration file**

```sh
- addresses:
  - address: localhost
    port: 80
  labels:
    app:label: app
```

Let's ignore the labels part for now; it is meant for more advanced configuration.
The `addresses` section specifies the registered destination addresses to which the policies should be applied.

Let's create a service registry entry for the echo-server service with the following command:

```sh
SVC_IP=$(kubectl get svc echo -o jsonpath='{@.spec.clusterIP}')
envsubst <<EOF | sudo tee /etc/camblet/services/echo-service.yaml >/dev/null
- addresses:
  - address: $SVC_IP
    port: 80
  labels:
    app:label: echo-server
EOF
```

### Try to connect to the echo server with certificate and configured Camblet

With the service registry entry for the echo-server service in place, let's check if its indeed fixed the error.
To do that, execute once again into the Alpine container and run cURL.

```sh
kubectl exec -it alpine -- curl echo
```

> > A second try could be necessary to achive the wanted effect!

Reviewer: ```suggestion
> A second attempt may be necessary to achieve the desired effect.
```
